starlette/routing.py

```diff
@@ -470,6 +483,54 @@ def __eq__(self, other: typing.Any) -> bool:
         )
 
 
+def _wrap_agen_lifespan_context(
+    lifespan_context: typing.Callable[[typing.Any], typing.AsyncGenerator]
+) -> typing.Callable[[typing.Any], typing.AsyncContextManager]:
+    @functools.wraps(lifespan_context)
+    @asynccontextmanager
+    async def agen_wrapper(
+        app: typing.Any,
+    ) -> typing.AsyncGenerator[None, None]:
+        async with aclosing(lifespan_context(app)) as agen:  # type: ignore
+            async for _ in agen:
+                yield
+
+    return agen_wrapper
+
+
+def _wrap_gen_lifespan_context(
+    lifespan_context: typing.Callable[[typing.Any], typing.Generator]
+) -> typing.Callable[[typing.Any], typing.AsyncContextManager]:
+    @functools.wraps(lifespan_context)
+    @asynccontextmanager
+    async def gen_wrapper(
+        app: typing.Any,
+    ) -> typing.AsyncGenerator[None, None]:
+        with contextlib.closing(lifespan_context(app)) as gen:
+            for _ in gen:
+                yield
+
+    return gen_wrapper
+
+
+def _wrap_lifespan_context(
```

Comment:
this stuff is a bit of a huge mess I think it's better to just make this a breaking change and make people add `contextlib.asynccontextmanager` to their functions

Reply:
I agree...

Reply:
@tomchristie are you happy for this to be a breaking change?

Reply:
Could you explain what would be necessary for a user to convert their existing lifespan generator to an `asynccontextmanager`? i.e. an example?

Reply:
You just add the `@contextlib.asynccontextmanager` decorator

Reply:
@JayH5 thank you so much for this comment! Because of this I worked out I could get rid of all my ugly _wrap_lifespan_context stuff and just use `asynccontextmanager(lifespan_context)`